Sources/OpenCombine/Publishers/Publishers.Filter.swift

```diff
@@ -0,0 +1,208 @@
+    public func filter(
+        _ isIncluded: @escaping (Self.Output) -> Bool
+    ) -> Publishers.Filter<Self> {
+        return Publishers.Filter(upstream: self, isIncluded: isIncluded)
+    }
+
+    /// Republishes all elements that match a provided error-throwing closure.
+    ///
+    /// If the `isIncluded` closure throws an error, the publisher fails with that error.
+    ///
+    /// - Parameter isIncluded:  A closure that takes one element and returns a
+    /// Boolean value indicating whether to republish the element.
+    /// - Returns:  A publisher that republishes all elements that satisfy the closure.
+    public func tryFilter(
+        _ isIncluded: @escaping (Self.Output) throws -> Bool
+    ) -> Publishers.TryFilter<Self> {
+        return Publishers.TryFilter(upstream: self, isIncluded: isIncluded)
+    }
+}
+
+private final class _Filter<Upstream: Publisher, Downstream: Subscriber>
+    : _BaseFilter<Upstream, Downstream>,
+      Subscriber,
+      CustomStringConvertible
+    where Upstream.Output == Downstream.Input,
+          Upstream.Failure == Downstream.Failure {
+
+    var description: String { return "Filter" }
+
+    init(downstream: Downstream, isIncluded: @escaping (Input) -> Bool) {
+        super.init(downstream: downstream,
+                   errorTransform: { $0 },
+                   isIncluded: { .success(isIncluded($0))
+        })
+    }
+}
+
+private final class _TryFilter<Upstream: Publisher, Downstream: Subscriber>
+    : _BaseFilter<Upstream, Downstream>,
+    Subscriber,
+    CustomStringConvertible
+    where Upstream.Output == Downstream.Input, Downstream.Failure == Error {
+
+    init(downstream: Downstream, isIncluded: @escaping (Input) throws -> Bool) {
+        super.init(downstream: downstream,
+                   errorTransform: { $0 as Error },
+                   isIncluded: { input in Result { try isIncluded(input) }
+        })
+    }
+
+    var description: String { return "TryFilter" }
+}
+
+private class _BaseFilter<Upstream: Publisher, Downstream: Subscriber>
+    : OperatorSubscription<Downstream>,
+      Subscription
+    where Upstream.Output == Downstream.Input
+{
+    typealias Input = Upstream.Output
+    typealias Output = Downstream.Input
+    typealias Failure = Upstream.Failure
+
+    private let _isIncluded: (Input) -> Result<Bool, Downstream.Failure>
+    private var _demand: Subscribers.Demand = .none
+
+    var errorTransform: (Upstream.Failure) -> Downstream.Failure
+
+    init(downstream: Downstream,
+         errorTransform: @escaping (Upstream.Failure) -> Downstream.Failure,
+         isIncluded: @escaping (Input) -> Result<Bool, Downstream.Failure>) {
+        self._isIncluded = isIncluded
+        self.errorTransform = errorTransform
+        super.init(downstream: downstream)
+    }
+
+    func receive(subscription: Subscription) {
+        upstreamSubscription = subscription
+        subscription.request(.unlimited)
```

Comment:
Can you please add a test to check if Apple's Combine also requests `.unlimited` from the upstream upon subscription?

Reply:
@broadwaylamb I’ve added a test and it passing for me locally in compatibility mode.

Reply:
@broadwaylamb I’ve added a test and it passing for me locally in compatibility mode.